Translate Chapter 19 to PTBR (#33)
JuliaIntro/JuliaIntroBR.jl: book/src/chap19.asciidoc

[[chap19]]
== Extra: Sintaxe

> Uma das metas deste livro é ensinar o mínimo possível de Julia. Quando havia duas formas de fazer algo, escolhíamos uma e evitavamos mencionar a outra. Ou às vezes, utilizamos a segunda forma como exercício.

Reviewer: ```suggestion
Uma das nossas metas para este livro é ensinar o mínimo possível de Julia. Quando havia duas formas de fazer algo, escolhíamos uma e evitávamos mencionar a outra. Ou às vezes, propúnhamos a segunda forma como exercício.
```

> Agora queremos retornar a algumas coisas boas que ficaram pra trás. O Julia oferece uma variadade de recursos que não são realmente necessários – você pode escrever um bom código sem eles – mas com eles é possível escrever um código mais conciso, legível ou eficiente e, às vezes, todos os três.

Reviewer: ```suggestion
Agora queremos retornar a algumas coisas boas que ficaram pra trás. O Julia oferece uma variadade de recursos que não são realmente necessários—você pode escrever um bom código sem eles—mas com eles pode-se escrever um código mais conciso, legível ou eficiente e, às vezes, todos os três.
```

> Neste capítulo e no próximo, iremos discutir coisas que deixamos para trás nos capítulos anteriores:

Reviewer: ```suggestion
Neste e no próximo capítulo, iremos discutir coisas dos capítulos anteriores que deixamos para trás:
```

> * suplementos de sintaxe

Reviewer: ```suggestion
* os suplementos de sintaxe
```

> * funções, tipos e macros diretamente disponíveis em +Base+

Reviewer: ```suggestion
* as funções, os tipos e as macros diretamente disponíveis em +Base+
```

(((Base)))

> * funções, tipos e macros da Biblioteca Padrão

Reviewer: ```suggestion
* as funções, os tipos e as macros da Biblioteca Padrão
```

(((Biblioteca Padrão)))

=== Tuplas Nomeadas

> Podemos nomear os componentes de uma tupla, criando uma tupla nomeada.

Reviewer: ```suggestion
Você pode nomear os componentes de uma tupla, criando uma tupla nomeada.
```

1
----

> Com tuplas nomeadas, os campos podem ser acessados por nome usando a sintaxe do ponto +(x.a)+.

Reviewer: ```suggestion
Nas tuplas nomeadas, os campos podem ser acessados pelo nome usando a sintaxe do ponto +(x.a)+.
```

> (((tuplas nomeadas)))(((sintaxe do ponto)))

Reviewer: ```suggestion
(((tupla nomeada)))(((sintaxe do ponto)))
```

=== Funções

> Uma função no Julia pode ser definida através de uma sintaxe compacta:

Reviewer: ```suggestion
Uma função no Julia pode ser definida por uma sintaxe compacta:
```

#3 (generic function with 1 method)
----

> Esses são exemplos de _funções anônimas_. Funções anônimas geralmente são usadas como argumentos para outras funções:

Reviewer: ```suggestion
Estes são os exemplos de _funções anônimas_. E elas geralmente são usadas como argumentos para outras funções:
```
Estou usando estes porque é um passado próximo.

> (((funções anônimas)))(((Plots)))((("módulo", "Plots", see="Plots")))(((plot)))((("função", "Plots", "plot", see="plot")))

Reviewer: ```suggestion
(((função anônima)))(((Plots)))((("módulo", "Plots", see="Plots")))(((plot)))((("função", "Plots", "plot", see="plot")))
```

----

<<fig19-1>> mostra a saída do comando plot.

[[fig19-1]]
.Plot
image::images/fig191.svg[pdfwidth="10cm"]

==== Argumentos de Palavras-Chave

Argumentos de funções também podem ser nomeadas:

[source,@julia-repl-test]
----
julia> function meu_plot(x, y; style="solid", width=1, color="black")
           ###
       end
meu_plot (generic function with 1 method)
julia> meu_plot(0:10, 0:10, style="dotted", color="blue")

----

_Argumentos de palavras-chave_ em uma função são especificados depois de um ponto e vírgula na assinatura, apesar de poder ser chamado com uma vírgula.
(((;)))(((argumentos de palavra-chave)))

==== Fechamentos

> Um _fechamento_ é uma técnica que permite uma função capturar uma variável definida fora do escopo da chamada de função.

Reviewer: ```suggestion
Um _fechamento_ é uma técnica que permite uma função capturar uma variável definida fora do escopo da chamada da função.
```

> 

Reviewer: Trocar **_myplot_** para **_meu_plot_** na linha 80, 83 e 84.

----

> _Argumentos de palavras-chave_ em uma função são especificadas depois de um ponto e vírgula na assinatura, mas pode ser chamado com uma vírgula.

Reviewer: ```suggestion
_Argumentos de palavras-chave_ em uma função são especificados depois de um ponto e vírgula na assinatura, apesar de poder ser chamado com uma vírgula.
```

(((;)))(((argumentos de palavra-chave)))

> === Fechamentos

Reviewer: ```suggestion
==== Fechamentos
```

> Um _fechamento_ é a técnica que permite uma função capturar uma variável definida fora do escopo de chamada de função.

Reviewer: ```suggestion
Um _fechamento_ é uma técnica que permite uma função capturar uma variável definida fora do escopo da chamada de função.
```

1
----

> Nesse exemplo, a função +foo+ retorna uma função anônima que possui acesso ao argumento +x+ da função +foo+. +bar+ aponta para a função anônima e retorna o valor do argumento de +foo+.

Reviewer: ```suggestion
Nesse exemplo, a função +foo+ retorna uma função anônima que possui acesso ao argumento +x+ da função +foo+. A variável +bar+ aponta para a função anônima e retorna o valor do argumento de +foo+.
```

=== Blocos

> Um _bloco_ é uma forma de agrupar diversos comandos. Um bloco começa com a palavra chave +begin+ e finaliza com +end+.

Reviewer: ```suggestion
Um _bloco_ é uma forma de agrupar diversos comandos. Um bloco começa com a palavra chave +begin+ e termina com +end+.
```

> Um _bloco_ é uma forma de agrupar diversos comandos. Um bloco começa com a palavra chave +begin+ e termina com +end+.

Reviewer: ```suggestion
Um _bloco_ é uma forma de agrupar diversos comandos. Um bloco começa com a palavra-chave +begin+ e termina com +end+.
```

(((begin)))((("palavra-chave", "begin", see="begin")))(((end)))(((bloco)))

> No <<chap04>> a macro +@svg+ foi introduzida:

Reviewer: ```suggestion
No <<chap04>>, a macro +@svg+ foi introduzida:
```

end
----

> Neste exemplo, a macro +@svg+ possui um argumento único, i.e. um bloco agrupando 3 chamadas de funções.

Reviewer: ```suggestion
Neste exemplo, a macro +@svg+ possui um argumento único, i.e. um bloco que está agrupando 3 chamadas de função.
```

==== Blocos +let+

> Um bloco +let+ é últil para criar novas ligações, i.e. locais que podem se referir a valores.

Reviewer: ```suggestion
Um bloco +let+ é útil para criar novas ligações, i.e. locais que podem se referir a valores.
```

z = -1
----

> No exemplo, a primeira macro +@show+ mostra o +x+ local, a global +y+ e a local indefinida +z+. As variáveis globais são intocáveis.

Reviewer: ```suggestion
Nesse exemplo, a primeira macro +@show+ mostra o +x+ local, a global +y+ e a local indefinida +z+. As variáveis globais são intocáveis.
```

> No exemplo, a primeira macro +@show+, mostra o +x+ local, a global +y+ e a local indefinida +z+. As variáveis globais são intocáveis.

Reviewer: ```suggestion
No exemplo, a primeira macro +@show+ mostra o +x+ local, a global +y+ e a local indefinida +z+. As variáveis globais são intocáveis.
```

==== Blocos +do+

> Em <<reading_and_writing>> tivemos que fechar o arquivo após o término da gravação. Isso pode ser feito automaticamente usando um _bloco do_:

Reviewer: ```suggestion
Em <<reading_and_writing>>, tivemos que fechar o arquivo após o término da escrita. Isso pode ser feito automaticamente usando um _bloco do_:
```

(((do)))((("palavra-chave", "do", see="do")))

[source,@julia-repl-test chap19]
----

> julia> dado = "Aqui está a acácia,\no emblema de nossa terra.\n"

Reviewer: Sugestão para colocar "algo brasileiro":
```suggestion
julia> dado = "O meu nome é Severino, \nnão tenho outro de pia.\n"
```

> "Aqui está a acácia,\no emblema de nossa terra.\n"

Reviewer: ```suggestion
"O meu nome é Severino, \nnão tenho outro de pia.\n"
```

julia> open("output.txt", "w") do fout
           write(fout, dado)
       end
48
----

> Nesse exemplo, +fout+ é um fluxo de arquivo usado para saída.

Reviewer: ```suggestion
Nesse exemplo, +fout+ é um fluxo de arquivo usado para a saída.
```

end
----

> Um bloco +do+ pode "capturar" variáveis de seus escopos anexos. Por exemplo, a variável +dado+ no exemplo acima de +open pass:[...] do+ é capturada de outro escopo.

Reviewer: ```suggestion
Um bloco +do+ pode “capturar” as variáveis de seus escopos anexos. Por exemplo, a variável +dado+ no exemplo acima de +open pass:[...] do+ é capturada de outro escopo.
```

=== Controle de Fluxo

==== Operador Ternário

O _operador ternário_, +?:+, é uma alternativa para uma declaração +if-else+ usada quando você precisa fazer uma escolha entre valores de expressão única.

> (((?:)))((("operator", "Base", "?:", see="?:")))((("operador ternário", see="?:")))

Reviewer: ```suggestion
(((?:)))((("operador", "Base", "?:", see="?:")))((("operador ternário", see="?:")))
```

end
----

> Um bloco +do+ pode “capturar” as variáveis de seus escopos anexos. Por exemplo, a variável +dado+ no exemplo acima de +open pass:[...] do+ é capturada de outro escopo.

Reviewer: ```suggestion
Um bloco +do+ pode “capturar” as variáveis dos seus escopos anexos. Por exemplo, a variável +dado+ em +open pass:[...] do+ no exemplo acima é capturada de outro escopo.
```

=== Controle de Fluxo

==== Operador Ternário

> O _operador ternário_, +?:+, é uma alternativa para uma declaração +if-else+ usada quando você precisa fazer uma escolha entre valores de expressão única.

Reviewer: ```suggestion
O _operador ternário_ +?:+ é uma alternativa para uma declaração +if-else+ usada quando você precisa fazer uma escolha entre valores de expressão única.
```

(((?:)))((("operador", "Base", "?:", see="?:")))((("operador ternário", see="?:")))

[source,@julia-repl-test]
----
julia> a = 150
150
julia> a % 2 == 0 ? println("par") : println("ímpar")
par
----

A expressão antes de +?+ é uma condição. Se a condição for +true+, a expressão antes de +:+ é avaliada, caso contrário, a expressão depois de +:+ é avaliada.

==== Avaliação de Curto-Circuíto

Os operadores +&&+ e +||+ fazem uma _avaliação de curto-circuito_: o próximo argumento só é avaliado quando é necessário determinar o valor final.
(((&&)))(((||)))(((avaliação de curto-circuito)))

Por exemplo, uma rotina recursiva do fatorial pode ser definida da seguinte maneira:
(((fat)))

[source,@julia-setup]
----
function fat(n::Integer)
    n >= 0 || error("n deve ser inteiro não negativo")
    n == 0 && return 1
    n * fat(n-1)
end
----

> ==== Tarefas (também conhecido como Corrotina)

Reviewer: ```suggestion
==== Tarefas (também conhecida como Corrotina)
```

> even

Reviewer: Traduzir a linha 210 e 211.

----

> A expressão antes de +?+ é uma expressão de condição. Se a condição for +true+, a expressão antes de +:+ é avaliada, caso contrário, a expressão depois de +:+ é avaliada.

Reviewer: ```suggestion
A expressão antes de +?+ é uma condição. Se a condição for +true+, a expressão antes de +:+ é avaliada, caso contrário, a expressão depois de +:+ é avaliada.
```

==== Avaliação de Curto-Circuíto

> Os operadores +&&+ e +||+ fazem uma _avaliação de curto-circuito_: O próximo argumento só é avaliado quando é necessário para determinar o valor final.

Reviewer: ```suggestion
Os operadores +&&+ e +||+ fazem uma _avaliação de curto-circuito_: o próximo argumento só é avaliado quando é necessário determinar o valor final.
```

(((&&)))(((||)))(((avaliação de curto-circuito)))

> Por exemplo, uma rotira de fatorial por recursão pode ser definida da seguinte maneira:

Reviewer: ```suggestion
Por exemplo, uma rotina recursiva do fatorial pode ser definida da seguinte maneira:
```

(((fat)))

[source,@julia-setup]
----
function fat(n::Integer)

>     n >= 0 || error("n deve ser inteiro positivo")

Reviewer: Na matemática, principalmente em otimização, há diferença entre não negativo e positivo. Não negativo inclui o zero e o positivo não inclui.
```suggestion
    n >= 0 || error("n deve ser inteiro não negativo")
```

    n == 0 && return 1
    n * fat(n-1)
end
----

> ==== Tarefas (também conhecido Corrotina)

Reviewer: ```suggestion
==== Tarefas (também conhecido como Corrotina)
```

> Uma _tarefa_ é uma estrutura de controle que pode passar pelo controle cooperativo sem retornar. Em Julia, uma tarefa pode ser implementada como uma função tendo como primeiro argumento um objeto de canal (+Channel+). Um canal é usado para passar valores de uma função ao chamador.

Reviewer: ```suggestion
Uma _tarefa_ é uma estrutura de controle que pode passar pelo controle cooperativamente sem retornar. Em Julia, uma tarefa pode ser implementada como uma função tendo como o primeiro argumento um objeto de canal (+Channel+). Um canal é usado para passar valores de uma função para quem a chamou.
```

> A sequência de Fibonnaci pode ser gerada usando uma tarefa.

Reviewer: ```suggestion
A sequência de Fibonnaci pode ser gerada por uma tarefa.
```

> (((tarefa)))(((Canal)))((("tipo", "Base", "Channel", see="Channel")))(((put!)))((("função", "Base", "put!", see="put!")))

Reviewer: ```suggestion
(((tarefa)))(((Channel)))((("tipo", "Base", "Channel", see="Channel")))(((put!)))((("função", "Base", "put!", see="put!")))
```

end
----

> +put+ armazena valores em um objeto de canal e +take!+ lê valores de:

Reviewer: ```suggestion
+put+ armazena os valores em um objeto de canal e +take!+ lê os valores dele:
```

3
----

> O construtor +Channel+ cria uma tarefa. A função +fib+ é suspendida depois de cada chamada para +put!+ e retorna depois de +take!+. Por questões de performance, muitos valores das sequências são armazenadas em buffer no objeto de canal durante um ciclo de retomada/suspensão.

Reviewer: ```suggestion
O construtor +Channel+ cria a tarefa. A função +fib+ é suspendida aṕos cada chamada para +put!+ e retorna depois de +take!+. Por questões de performance, diversos valores das sequências são armazenadas temporariamente no objeto de canal durante um ciclo de retomada/suspensão.
```

> O construtor +Channel+ cria a tarefa. A função +fib+ é suspendida aṕos cada chamada para +put!+ e retorna depois de +take!+. Por questões de performance, diversos valores das sequências são armazenadas temporariamente no objeto de canal durante um ciclo de retomada/suspensão.

Reviewer: ```suggestion
O construtor +Channel+ cria a tarefa. A função +fib+ é suspendida aṕos cada chamada para +put!+ e retorna depois de +take!+. Por questões de performance, diversos valores da sequência são armazenados temporariamente no objeto de canal durante um ciclo de retomada/suspensão.
```

----

=== Tipos

==== Tipos Primitivos

> Um tipo concreto que consiste em bits antigos simples é chamado de _tipo primitivo_. Ao contrário da maioria das linguagens, com Julia, você pode declarar seus próprios tipos primitivos. Os tipos primitivos padrão são definidos da mesma maneira:

Reviewer: ```suggestion
Um tipo concreto que consiste em bits simples e antigos é chamado de _tipo primitivo_. Ao contrário da maioria das linguagens, no Julia, você pode declarar seus próprios tipos primitivos. Os tipos primitivos padrões são definidos da mesma maneira:
```

> Um tipo concreto que consiste em bits simples e antigos é chamado de _tipo primitivo_. Ao contrário da maioria das linguagens, no Julia, você pode declarar seus próprios tipos primitivos. Os tipos primitivos padrões são definidos da mesma maneira:

Reviewer: ```suggestion
Um tipo concreto que consiste em bits simples e antigos é chamado de _tipo primitivo_. Ao contrário da maioria das linguagens, no Julia você pode declarar os seus próprios tipos primitivos. Os tipos primitivos padrões são definidos da mesma maneira:
```

primitive type Int64 <: Signed 64 end
----

O número nos comandos especificam quantos bits são necessários.

> O exemplo à seguir cria um tipo primitivo +Byte+ e um construtor:

Reviewer: ```suggestion
O exemplo a seguir cria um tipo primitivo +Byte+ e um construtor:
```

Byte(0x01)
----

> A função +reinterpret+ é usada para armazenar os bits de um inteiro não assinado com 8 bits (+UInt8+) no bit.

Reviewer: ```suggestion
A função +reinterpret+ é usada para armazenar os bits de um inteiro não assinado com 8 bits (+UInt8+) no byte.
```

> (((reinterpret)))((("função", "Base", "reinterpret", see="reinterpret")))(((UInt8)))((("Tipo", "Base", "UInt8", see="UInt8")))

Reviewer: ```suggestion
(((reinterpret)))((("função", "Base", "reinterpret", see="reinterpret")))(((UInt8)))((("tipo", "Base", "UInt8", see="UInt8")))
```

==== Tipos Paramétricos

> O tipo de sistema do Julia é _paramétrico_, o que significa que os tipos podem possuir parâmetros.

Reviewer: ```suggestion
O tipo de sistema do Julia é _paramétrico_, significando que os tipos podem possuir parâmetros.
```

> Tipos de Parâmetros são introduzidos depois do nome do tipo, cercado por chaves:

Reviewer: ```suggestion
Parâmetros do tipo são introduzidos depois do nome do tipo, cercado por chaves:
```

Byte(0x01)
----

A função +reinterpret+ é usada para armazenar os bits de um inteiro não assinado com 8 bits (+UInt8+) no byte.
(((reinterpret)))((("função", "Base", "reinterpret", see="reinterpret")))(((UInt8)))((("tipo", "Base", "UInt8", see="UInt8")))

==== Tipos Paramétricos

O tipo de sistema do Julia é _paramétrico_, significando que os tipos podem possuir parâmetros.

> Parâmetros do tipo são introduzidos depois do nome do tipo, cercado por chaves:

Reviewer: ```suggestion
Parâmetros de tipo são introduzidos depois do nome do tipo, cercado por chaves:
```

(((chaves)))

[source,@julia-setup chap19]
----
struct Ponto{T<:Real}
    x::T
    y::T
end
----

Isso define um novo tipo paramétrico, +Ponto{T<:Real}+, segurando duas "coordenadas" do tipo +T+, da qual pode ser de qualquer tipo desde que tenha +Real+ como supertipo.

[source,@julia-repl-test chap19]
----
julia> Ponto(0.0, 0.0)
Ponto{Float64}(0.0, 0.0)
----

> Além dos tipos compostos, tipos abstratos e tipos primitivos também podem ter um parâmetro do tipo.

Reviewer: ```suggestion
Além dos tipos compostos, tipos abstratos e tipos primitivos também podem ter um parâmetro de tipo.
```

[TIP]
====
Ter tipos concretos nos campos da struct é absolutamente recomendado por motivos de desempenho, portanto essa é uma boa maneira de tornar +Ponto+ rápido e flexível.
====

> ==== União dos Tipos

Reviewer: ```suggestion
==== União de Tipo
```

> end

Reviewer: Traduzir **_Point_** para **_Ponto_** na linha 328, 338 e 339.

----

> Isso define um novo tipo paramétrico, +Point{T<:Real}+, segurando duas "coordenadas" de tipo +T+, da qual pode ser qualquer tipo contendo +Real+ como supertipo.

Reviewer: ```suggestion
Isso define um novo tipo paramétrico, +Ponto{T<:Real}+, segurando duas "coordenadas" do tipo +T+, da qual pode ser de qualquer tipo desde que tenha +Real+ como supertipo.
```

[source,@julia-repl-test chap19]
----
julia> Point(0.0, 0.0)
Point{Float64}(0.0, 0.0)
----

> Além dos tipos compostos, tipos abstratos e tipos primitivos também podem ter um parâmetro de tipo.

Reviewer: ```suggestion
Além dos tipos compostos, tipos abstratos e tipos primitivos também podem ter um parâmetro do tipo.
```

[TIP]
====

> Ter tipos concretos para campos de estrutura é absolutamente recomendado por motivos de desempenho, portanto, é uma boa maneira de tornar o +Point+ rápido e flexível.

Reviewer: ```suggestion
Ter tipos concretos nos campos da struct é absolutamente recomendado por motivos de desempenho, portanto essa é uma boa maneira de tornar +Ponto+ rápido e flexível.
```

====

> ==== Tipo de União

Reviewer: ```suggestion
==== União dos Tipos
```

> Um _tipo de união_ é um tipo paramétrico abstrato que pode atuar como qualquer um dos seus tipos de argumento:

Reviewer: ```suggestion
Uma _união de tipo_ é um tipo paramétrico abstrato que pode atuar como qualquer um dos tipos do seu argumento:
```

> (((tipo união)))(((Union)))((("tipo", "Base", "Union", see="Union")))

Reviewer: ```suggestion
(((união de tipo)))(((Union)))((("tipo", "Base", "Union", see="Union")))
```

"Julia"
----

> Uma união de tipo é na maioria das linguagens de computador uma construção interna para raciocinar sobre tipos. Julia, no entanto, expõe esse recurso a seus usuários porque um código eficiente pode ser gerado quando a união de tipos possui um pequeno número de tipos. Esse recurso oferece ao programador Julia uma tremenda flexibilidade para controlar o envio.

Reviewer: ```suggestion
Uma união de tipo é na maioria das linguagens de computador uma construção interna para refletir sobre os tipos. O Julia, no entanto, expõe esse recurso aos seus usuários porque um código eficiente pode ser gerado quando a união de tipos possui um pequeno número de tipos. Esse recurso oferece ao programador do Julia uma tremenda flexibilidade para controlar o despacho.
```

"Julia"

> ----

Reviewer: Traduzir **_IntOrString_** nas linhas 356, 358 e 360.

> Uma união de tipo é na maioria das linguagens de computador uma construção interna para refletir sobre os tipos. O Julia, no entanto, expõe esse recurso aos seus usuários porque um código eficiente pode ser gerado quando a união de tipos possui um pequeno número de tipos. Esse recurso oferece ao programador do Julia uma tremenda flexibilidade para controlar o despacho.

Reviewer: ```suggestion
Uma união de tipo é na maioria das linguagens de computador uma construção interna para pensar sobre os tipos. O Julia, no entanto, expõe esse recurso aos seus usuários porque um código eficiente pode ser gerado quando a união de tipo possui um pequeno número de tipos. Esse recurso oferece ao programador do Julia uma tremenda flexibilidade para controlar o despacho.
```

true

> ----

Reviewer: Traduzir **_isintpoint_** na linha 375, 376 e 379, e traduzir **_Point_** na linha 377

==== Objetos Semelhantes a Funções

> Qualquer objeto arbitrário de Julia pode ser "chamado". Tais objetos "chamáveis" às vezes são chamados de "funtores".

Reviewer: ```suggestion
Qualquer objeto arbitrário de Julia pode ser “chamado”. Tais objetos “chamáveis” às vezes são denominados  _funtores_.
```

end

> ----

Reviewer: Verificar as traduções que devem ser feitas da linha 390 a linha 400. Eu sei q dá p traduzir Polynomial, mas não cheguei a olhar com mais atenção.

Author: Acredito que sim pois é uma Struct definida pelo programador.

931

> ----

Reviewer: Verificar as traduções que devem ser feitas da linha 407 e 408.

ERROR: MethodError: no method matching Point(::Int64, ::Float64)

> ----

Reviewer: Traduzir **_Point_** da linha 419 até 424, **_implicit_** nas linhas 419 e 423 e **_explicit_** na linha 421.

> Construtores internos e externos padrão são gerados para cada +T+:

Reviewer: ```suggestion
Construtores internos e externos padrões são gerados para cada +T+:
```

Point(x::T, y::T) where {T<:Real} = Point{T}(x,y);

> ----

Reviewer: Traduzir **_Point_** nas linhas 432, 435 e 438.

e ambos +x+ e +y+ devem ser do mesmo tipo.

> Quando +x+ e +y+ possuem tipos diferentes, o seguinte construtor externo pode ser definido:

Reviewer: ```suggestion
Quando +x+ e +y+ possuem tipos diferentes, o subsequente construtor externo pode ser definido como:
```

[source,@julia-setup chap19]
----

> Point(x::Real, y::Real) = Point(promote(x,y)...);

Reviewer: ```suggestion
Ponto(x::Real, y::Real) = Ponto(promote(x,y)...);
```

----

A função +promote+ é detalhada em <<promotion>>.
(((promote)))((("função", "Base", "promote", see="promote")))

=== Conversão e Promoção

> O Julia tem um sistema para promover argumentos para um tipo comum. Isso não é feito automaticamente, mas pode ser facilmente estendido.

Reviewer: ```suggestion
O Julia tem um sistema para promover argumentos para um tipo comum. Isso não é feito automaticamente, mas pode ser facilmente realizado.
```

==== Conversão

> Um valor pode ser convertido de um tipo para outro:

Reviewer: ```suggestion
Um valor pode ser convertido de um tipo para o outro:
```

UInt8
----

> Podemos adicionar nossos próprios métodos +convert+:

Reviewer: ```suggestion
Podemos adicionar os nossos próprios métodos +convert+:
```

> ----

Reviewer: Traduzir **_Point_** na linha 477 até 480.

[[promotion]]
==== Promoção

> _Promoção_ é a conversão de valores de tipos mistos para um único tipo comum:

Reviewer: ```suggestion
_Promoção_ é a conversão de valores dos tipos mistos para um único tipo comum:
```

(((promoção)))(((promote)))

[source,@julia-repl-test]
----
julia> promote(1, 2.5, 3)
(1.0, 2.5, 3.0)
----

> Os métodos para a função +promote+ normalmente não são definidos diretamente, mas a função auxiliar +promover_rule+ é usada para especificar as regras da promoção:

Reviewer: ```suggestion
Em geral, os métodos para a função +promote+ não são diretamente definidos, mas a função auxiliar +promote_rule+ é usada para especificar as regras da promoção:
```

> (((promote_rule)))((("function", "Base", "promote_rule", see="promote_rule")))

Reviewer: ```suggestion
(((promote_rule)))((("função", "Base", "promote_rule", see="promote_rule")))
```

[source,julia]
----
promote_rule(::Type{Float64}, ::Type{Int32}) = Float64
----

> === Meta-programação

Reviewer: ```suggestion
=== Metaprogramação
```

> O código Julia pode ser representado como uma estrutura de dados da própria linguagem. Isso permite que um programa transforme e gere seu próprio código.

Reviewer: ```suggestion
O código Julia pode ser representado como uma estrutura de dados da própria linguagem. Isso permite que um programa transforme e gere o seu próprio código.
```

==== Expressões

Todo programa do Julia começa como uma string:

[source,@julia-repl-test chap19]
----
julia> prog = "1 + 2"
"1 + 2"
----

> A próxima etapa é analisar cada string em um objeto chamado _expression_, representado pelo tipo Julia +Expr+:

Reviewer: ```suggestion
A próxima etapa é analisar cada string em um objeto chamado _expression_, representado pelo tipo +Expr+ do Julia:
```

> (((expressão)))(((Expr)))((("tipo", "Base", "Expr", see="Expr")))(((parse)))((("função", "Meta", "parse", see="parse")))

Reviewer: ```suggestion
(((expression)))(((Expr)))((("tipo", "Base", "Expr", see="Expr")))(((parse)))((("função", "Meta", "parse", see="parse")))
```

    3: Int64 2
----

> A função +dump+ exibe objetos expr com anotações.

Reviewer: ```suggestion
A função +dump+ detalha os objetos expr.
```

(((dump)))

> Expressões podem ser construídas diretamente prefixando +:+ entre parênteses ou usando um bloco de aspas

Reviewer: ```suggestion
Expressões podem ser diretamente construídas prefixando +:+ entre parênteses ou usando uma citação em bloco
```

3
----

> Cada módulo possui sua própria função +eval+ que avalia expressões em seu escopo.

Reviewer: ```suggestion
Cada módulo possui sua própria função +eval+ que avalia as expressões em seu escopo.
```

[WARNING]
====

> Quando você está usando muitas chamadas para a função +eval+, geralmente isso significa que algo está errado. +eval+ é considerado "mal".

Reviewer: ```suggestion
Quando você está usando muitas chamadas para a função +eval+, geralmente isso significa que algo está errado. A função +eval+ é considerado do “mal”.
```

====

==== Macros

> Macros podem incluir código gerado em um programa. Um _macro_ mapeia uma tupla de objetos +Expr+ diretamente para uma expressão compilada:

Reviewer: ```suggestion
Macros podem incluir o código gerado em um programa. Uma _macro_ mapeia uma tupla de objetos +Expr+ diretamente para uma expressão compilada:
```

(((macro)))

Aqui está uma simples macro:
(((@containervariavel)))((("macro", "definido pelo programador", "@containervariavel", see="@containervariavel")))

[source,@julia-setup chap19]
----
macro containervariavel(container, elemento)
    return esc(:($(Symbol(container,elemento)) = $container[$elemento]))
end
----

> As macros são chamadas com o prefixo +@+ (sinal de arroba). A chamada de macro +@containervariavel letters 1+ é substituída por:

Reviewer: ```suggestion
As macros são chamadas ao colocar o prefixo +@+ (sinal de arroba) em seus nomes. A macro chamada +@containervariavel letras 1+ é substituída por:
```

(((@)))

[source,julia]
----

> :(letters1 = letters[1])

Reviewer: Traduzir **_letters_**.

----

> +@macroexpand @containervariavel letters 1+ retorna essa expressão que é extremamente útil para depuração.

Reviewer: ```suggestion
+@macroexpand @containervariavel letras 1+ retorna essa expressão que é extremamente útil para a depuração.
```

(((@macroexpand)))((("macro", "Base", "@macroexpand", see="@macroexpand")))

> Este exemplo ilustra como uma macro pode acessar o nome de seus argumentos, algo que uma função não pode fazer. A expressão de retorno precisa ser "escapada" com +esc+, pois precisa ser resolvida no ambiente de chamada de macro.

Reviewer: ```suggestion
Este exemplo ilustra como uma macro pode acessar o nome dos seus argumentos, algo que uma função não pode fazer. O comando return precisa ser “escapada” com +esc+, pois precisa ser resolvida no ambiente da chamada da macro.
```

(((esc)))((("função", "Base", "esc", see="esc")))

[NOTE]
====
Por que macros?

> As macros geram e incluem fragmentos de código personalizado durante o tempo de análise, portanto, _antes_ que o programa completo seja executado.

Reviewer: ```suggestion
As macros geram e incluem fragmentos de código personalizado durante o tempo de análise, portanto, _antes_ da execução do programa completo.
```

====

==== Funções Geradas

> A macro +@generated+ cria código especializado para métodos, dependendo dos tipos de argumentos:

Reviewer: ```suggestion
A macro +@generated+ cria código especializado para os métodos, dependendo dos tipos de argumentos:
```

3
----

Cada módulo possui sua própria função +eval+ que avalia as expressões em seu escopo.

[WARNING]
====

> Quando você está usando muitas chamadas para a função +eval+, geralmente isso significa que algo está errado. A função +eval+ é considerado do “mal”.

Reviewer: ```suggestion
Quando você está usando muitas chamadas para a função +eval+, geralmente isso significa que algo está errado. A função +eval+ é considerada do “mal”.
```

====

==== Macros

Macros podem incluir o código gerado em um programa. Uma _macro_ mapeia uma tupla de objetos +Expr+ diretamente para uma expressão compilada:
(((macro)))

> Aqui está uma simples macro:

Reviewer: ```suggestion
Aqui está uma macro simples:
```

> (((@containervariavel)))((("macro", "definido pelo programador", "@containervariavel", see="@containervariavel")))

Reviewer: ```suggestion
(((@containervariável)))((("macro", "definido pelo programador", "@containervariável", see="@containervariável")))
```

[source,@julia-setup chap19]
----

> macro containervariavel(container, elemento)

Reviewer: ```suggestion
macro containervariável(container, elemento)
```

    return esc(:($(Symbol(container,elemento)) = $container[$elemento]))
end
----

> As macros são chamadas ao colocar o prefixo +@+ (sinal de arroba) em seus nomes. A macro chamada +@containervariavel letras 1+ é substituída por:

Reviewer: ```suggestion
As macros são chamadas ao colocar o prefixo +@+ (sinal de arroba) em seus nomes. A macro chamada +@containervariável letras 1+ é substituída por:
```

(((@)))

[source,julia]
----
:(letras1 = letras[1])
----

> +@macroexpand @containervariavel letras 1+ retorna essa expressão que é extremamente útil para a depuração.

Reviewer: ```suggestion
+@macroexpand @containervariável letras 1+ retorna essa expressão que é extremamente útil para a depuração.
```

(((@macroexpand)))((("macro", "Base", "@macroexpand", see="@macroexpand")))

> Este exemplo ilustra como uma macro pode acessar o nome dos seus argumentos, algo que uma função não pode fazer. O comando return precisa ser “escapada” com +esc+, pois precisa ser resolvida no ambiente da chamada da macro.

Reviewer: ```suggestion
Este exemplo ilustra como uma macro pode acessar o nome dos seus argumentos, algo que uma função não pode fazer. O comando return precisa ser “escapado” com +esc+, pois precisa ser resolvido no ambiente da chamada da macro.
```

(((esc)))((("função", "Base", "esc", see="esc")))

[NOTE]
====
Por que macros?

As macros geram e incluem fragmentos de código personalizado durante o tempo de análise, portanto, _antes_ da execução do programa completo.
====

==== Funções Geradas

> A macro +@generated+ cria código especializado para os métodos, dependendo dos tipos de argumentos:

Reviewer: ```suggestion
A macro +@generated+ cria código especializado para os métodos, dependendo dos tipos dos argumentos:
```

(((funções geradas)))(((@generated)))((("macro", "Base", "@generated", see="@generated")))

[source,@julia-setup chap19]
----
@generated function quadrado(x)
    println(x)
    :(x * x)
end
----

O corpo retorna uma expressão entre aspas como uma macro.

Para quem chama, a _função gerada_ se comporta como uma função regular:

[source,@julia-repl-test chap19]
----

> julia> x = quadrado(2); # nota: output é da declaração println() no corpo

Reviewer: ```suggestion
julia> x = quadrado(2); # nota: a saída é da declaração println() no corpo
```

end

> ----

Reviewer: Traduzir **_square_** na linha 614.

O corpo retorna uma expressão entre aspas como uma macro.

> Para o chamador, a função _generated_ se comporta como uma função regular:

Reviewer: ```suggestion
Para quem chama, a _função gerada_ se comporta como uma função regular:
```

[source,@julia-repl-test chap19]
----

> julia> x = square(2); # note: output é da declaração println() no corpo

Reviewer: ```suggestion
julia> x = quadrado(2); # nota: output é da declaração println() no corpo
```

Int64
julia> x              # agora imprimimos x
4

> julia> y = square("spam");

Reviewer: ```suggestion
julia> y = quadrado("spam");
```

  missing
----

> O tipo de elemento dessa lista é +Union{Missing, T}+, com +T+ o tipo de valores não ausentes.

Reviewer: ```suggestion
O tipo de elemento dessa lista é +Union{Missing, T}+, sendo +T+ o tipo de valores não ausentes.
```

> As funções de redução retornam +missing+ quando chamadas em listas que contêm valores ausentes

Reviewer: ```suggestion
As funções de redução retornam +missing+ quando chamadas nas listas que contêm valores ausentes
```

[source,@julia-repl-test chap19]
----
julia> sum(a)
missing
----

> Nessa situação, use a função +skipmissing+ para pular os valores ausentes:

Reviewer: ```suggestion
Nessa situação, use a função +skipmissing+ para ignorar os valores ausentes:
```

----

=== Chamando Códigos em C e Fortran

> Muitos códigos estão escritos em C ou Fortran. Reutilizar o código testado geralmente é melhor do que escrever sua própria versão de um algoritmo. Julia pode chamar diretamente as bibliotecas C ou Fortran existentes usando a sintaxe +ccall+.

Reviewer: ```suggestion
Muitos códigos estão escritos em C ou Fortran. Reutilizar o código testado geralmente é melhor do que escrever sua própria versão de um algoritmo. Julia pode chamar diretamente as bibliotecas existentes em C ou Fortran usando a sintaxe +ccall+.
```

(((ccall)))((("função", "Base", "ccall", see="ccall")))

> Em <<databases>>, introduzimos uma interface Julia na biblioteca GDBM de funções de banco de dados. A biblioteca está escrita em C. Para fechar o banco de dados, uma chamada de função para +close(db)+ deve ser feita:

Reviewer: ```suggestion
Em <<databases>>, introduzimos uma interface Julia para a biblioteca GDBM de funções de banco de dados. A biblioteca está escrita em C. Para fechar o banco de dados, uma chamada de função para +close(db)+ teve que ser feita:
```

end
----

> Um objeto dbm possui um campo +handle+ do tipo +Ptr{Cvoid}+. Este campo contém um ponteiro C que se refere ao banco de dados. Para fechar o banco de dados, a função C +gdbm_close+ deve ser chamada tendo como único argumento o ponteiro C apontando para o banco de dados e sem valor de retorno. O Julia faz isso diretamente com a função +ccall+ tendo como argumentos:

Reviewer: ```suggestion
Um objeto dbm possui um campo +handle+ do tipo +Ptr{Cvoid}+. Este campo guarda um ponteiro C que se refere ao banco de dados. Para fechar o banco de dados, a função C +gdbm_close+ deve ser chamada tendo como único argumento o ponteiro C apontando para o banco de dados e sem valor de retorno. O Julia faz isso diretamente com a função +ccall+ tendo como argumentos:
```

(((Ptr)))((("tipo", "Base", "Ptr", see="Ptr")))

> * uma trupla que consiste em um símbolo que contém o nome da função que queremos chamar: +:gdbm_close+ e a biblioteca compartilhada especificada como uma sequência: +"libgdm"+,

Reviewer: ```suggestion
* uma tupla que consiste em um símbolo que contém o nome da função que queremos chamar: +:gdbm_close+ e a biblioteca compartilhada especificada como uma string: +"libgdm"+,
```

* o tipo de retorno: +Cvoid+,

> * uma tupla com tipos de argumentos: +(Ptr{Cvoid},)+ e

Reviewer: ```suggestion
* uma tupla com os tipos de argumentos: +(Ptr{Cvoid},)+ e
```

> * uma tupla de tipos de argumentos: +(Ptr{Cvoid},)+ e

Reviewer: ```suggestion
* uma tupla com tipos de argumentos: +(Ptr{Cvoid},)+ e
```

* os valores do argumento: +handle+.

> O mapeamento completo da biblioteca GDBM pode ser encontrada como um exmplo nas fontes do JuliaIntroBR.

Reviewer: ```suggestion
O mapeamento completo da biblioteca GDBM pode ser encontrada como um exemplo no repositório do JuliaIntroBR.
```

> O mapeamento completo da biblioteca GDBM pode ser encontrada como um exemplo no repositório do JuliaIntroBR.

Reviewer: ```suggestion
O mapeamento completo da biblioteca GDBM pode ser encontrado como um exemplo no repositório do JuliaIntroBR.
```

=== Glossário

> fechamentos::

Reviewer: ```suggestion
fechamento::
```

> Função que captura variáveis de seu escopo definido.

Reviewer: ```suggestion
Função que captura as variáveis do seu escopo definido.
```

(((fechamento)))

> bloqueio let::

Reviewer: Na linha 129, vc traduz **_let block_** como **_bloco let_**. Gostaria de saber se vc vai adotar **_bloco let_** ou **_bloqueio let_** para a tradução de **_let block_**.

Depois de vc se decidir, se necessário, fazer mudança na linha 707 e 708.

> Bloqueia a alocação de novas ligações de variáveis.

Reviewer: ```suggestion
Bloco que aloca novas ligações de variáveis.
```

(((bloqueio let)))

> funções anônimas::

Reviewer: ```suggestion
função anônima::
```

> Função definida sem possuir nome.

Reviewer: ```suggestion
Função definida sem ter um nome.
```

(((função anônima)))

tupla nomeada::
Tupla com componentes nomeados.
(((tupla nomeada)))

argumentos de palavra-chave::

> Argumentos identificados pelo nome em vez de apenas pela posição.

Reviewer: ```suggestion
Argumentos identificados pelo nome e não só pela posição.
```

(((argumentos de palavra-chave)))

bloco do::
Construção de sintaxe usada para definir e chamar uma função anônima que se parece com um bloco de código normal.
(((bloco do)))

operador ternário::

> Operador de fluxo de controle usando três operandos para especificar uma condição, uma expressão a ser executada quando a condição render +true+ e uma expressão a ser executada quando a condição render +false+.

Reviewer: ```suggestion
Operador de fluxo de controle que usa três operandos para especificar uma condição, uma expressão a ser executada quando a condição produz +true+ e uma expressão a ser executada quando a condição produz +false+.
```

(((operador ternário)))

avaliação de curto-circuíto::

> Avaliação de um operador booleano para o qual o segundo argumento é executado ou avaliado apenas se o primeiro argumento não for suficiente para determinar o valor da expressão.

Reviewer: ```suggestion
Avaliação de um operador booleano na qual o segundo argumento é executado ou avaliado apenas se o primeiro argumento não for suficiente para determinar o valor da expressão.
```

(((avaliação de curto-circuíto)))

tarefa (também conhecida como Corrotina)::

> Recurso de fluxo de controle que permite a suspensão dos cálculos e a retomada deles de maneira flexível.

Reviewer: ```suggestion
Recurso de fluxo de controle que permite a suspensão e a retomada dos cálculos de maneira flexível.
```

> Recurso de fluxo de controle que permite que os cálculos sejam suspensos e retomados de maneira flexível.

Reviewer: ```suggestion
Recurso de fluxo de controle que permite a suspensão dos cálculos e a retomada deles de maneira flexível.
```

(((tarefa)))

tipo primitivo::

> Tipo de concreto cujos dados consistem em bits antigos simples.

Reviewer: ```suggestion
Tipo concreto cujos dados consistem em bits simples e antigos.
```

(((tipo primitivo)))

> tipo de união::

Reviewer: ```suggestion
união de tipo::
```

Tipo que inclui como objetos todas as instâncias de qualquer um dos seus parâmetros de tipo.

> (((tipo de união)))

Reviewer: ```suggestion
(((união de tipo)))
```

tipo paramétrico::

> Tipo que está parametrizado.

Reviewer: ```suggestion
Tipo que é parametrizado.
```

(((tipo paramétrico)))

funtor::
Objeto com um método associado, para que ele possa ser chamado.

> (((functor)))

Reviewer: ```suggestion
(((funtor)))
```

conversão::
A conversão permite converter um valor de um tipo para outro.
(((conversão)))

promoção::

> Conversão valores de tipos mistos em um único tipo comum.

Reviewer: ```suggestion
Conversão de valores de tipos mistos em um único tipo comum.
```

(((promoção)))

expressão::

> Tipo de Julia que contém uma construção de linguagem.

Reviewer: ```suggestion
Tipo do Julia que contém uma construção de linguagem.
```

(((expressão)))

macro::

> Forma de incluir o código gerado no corpo final de um programa.

Reviewer: ```suggestion
Forma de incluir o código gerado no final do corpo de um programa.
```

(((macro)))

funções geradas::
Funções capazes de gerar código especializado, dependendo dos tipos dos argumentos.
(((funções geradas)))

valores ausentes::

> Instâncias que representam pontos de dados sem valor.

Reviewer: ```suggestion
Instâncias que representam informações sem valor.
```
